src/server/pfs/server/master.go

```diff
@@ -139,6 +157,119 @@ func (d *driver) finishCommits(ctx context.Context) (retErr error) {
 	)
 }
 
+func (d *driver) manageBranches(ctx context.Context, repoKey string) error {
+	ctx, cancel := pctx.WithCancel(ctx)
+	defer cancel()
+	type cronTrigger struct {
+		cancel context.CancelFunc
+		spec   string
+	}
+	cronTriggers := make(map[string]*cronTrigger)
+	defer func() {
+		for _, ct := range cronTriggers {
+			ct.cancel()
+		}
+	}()
+	return d.branches.ReadOnly(ctx).WatchByIndexF(pfsdb.BranchesRepoIndex, repoKey, func(ev *watch.Event) error {
+		if ev.Type == watch.EventError {
+			return ev.Err
+		}
+		key := string(ev.Key)
+		if ev.Type == watch.EventDelete {
+			if ct, ok := cronTriggers[key]; ok {
+				ct.cancel()
+			}
+			delete(cronTriggers, key)
+			return nil
+		}
+		branchInfo, err := d.inspectBranch(ctx, pfsdb.ParseBranch(key))
+		if err != nil {
+			return err
+		}
+		if branchInfo.Trigger == nil || branchInfo.Trigger.CronSpec == "" {
+			return nil
+		}
+		// Only create a new goroutine if one doesn't already exist or the spec changed.
+		if ct, ok := cronTriggers[key]; ok {
+			if ct.spec == branchInfo.Trigger.CronSpec {
+				return nil
+			}
+			ct.cancel()
+			delete(cronTriggers, key)
+		}
+		ctx, cancel := pctx.WithCancel(ctx)
+		cronTriggers[key] = &cronTrigger{
+			cancel: cancel,
+			spec:   branchInfo.Trigger.CronSpec,
+		}
+		go func() {
+			backoff.RetryUntilCancel(ctx, func() error { //nolint:errcheck
+				return d.runCronTrigger(ctx, branchInfo.Branch)
+			}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
+				log.Error(ctx, "error running cron trigger", zap.String("branch", key), zap.Error(err), zap.Duration("retryAfter", d))
+				return nil
+			})
+		}()
+		return nil
+	})
+}
+
+func (d *driver) runCronTrigger(ctx context.Context, branch *pfs.Branch) error {
+	branchInfo, err := d.inspectBranch(ctx, branch)
```

Comment:
Should we be concerned about doing a read here?  We actually have access to the branch info in the caller, I think we should just pass that in directly.

Reply:
The branch can be deleted at any time. If it is deleted while this method is running, the method will return an error and be retried by the caller. If we don't reload the branch info after a retry, we will get stuck in a retry loop until the delete event is received. When we detect that the branch doesn't exist, we just log it and return nil since there is nothing to do.

Reply:
Should we check for the not exist error in the Retry's error handler?  That way we catch it where ever it occurs, instead of just at the start of this function?

Reply:
I considered putting it there, but I leaned towards the `runCronTrigger` method being responsible for loading the current branch info. It also seemed kind of nice for when it is eventually unit testable. But, I am reconsidering just putting it in the notify callback. That is the pattern we use in the worker registry and it may be better for `runCronTrigger` to operate on a stable branch info and always return an error if the branch is deleted.